contracts/registries/LicenseRegistryV2.sol

```diff
@@ -0,0 +1,382 @@
+        });
+        emit MintingLicenseSpecSetForAll(ipId, mintingLicenseSpec);
+    }
+
+    function attachLicenseConfigToIp(
+        address ipId,
+        address licenseTemplate,
+        uint256 licenseConfigId
+    ) external onlyLicensingModule {
+        if (!_existsLicenseConfig(licenseTemplate, licenseConfigId)) {
+            revert Errors.LicensingModule__LicenseConfigNotFound(licenseTemplate, licenseConfigId);
+        }
+
+        if (_isDerivativeIp(ipId)) {
+            revert Errors.LicensingModule__DerivativesCannotAddLicenseConfig();
+        }
+
+        LicenseRegistryStorage storage $ = _getLicenseRegistryStorage();
+        if ($.expireTimes[ipId] < block.timestamp) {
+            revert Errors.LicenseRegistry__IpExpired(ipId);
+        }
+        $.attachedLicenseConfigs[ipId].add(licenseConfigId);
+    }
+
+    // solhint-disable-next-line code-complexity
+    function registerDerivativeIp(
+        address derivativeIpId,
+        address[] calldata originalIpIds,
+        address licenseTemplate,
+        uint256[] calldata licenseConfigIds
+    ) external onlyLicensingModule {
+        if (originalIpIds.length == 0) {
+            revert Errors.LicenseRegistry__NoOriginalIp();
+        }
+        LicenseRegistryStorage storage $ = _getLicenseRegistryStorage();
+        if ($.attachedLicenseConfigs[derivativeIpId].length() > 0) {
+            revert Errors.LicenseRegistry__DerivativeIpAlreadyHasLicense(derivativeIpId);
+        }
+        if ($.originalIps[derivativeIpId].length() > 0) {
+            revert Errors.LicenseRegistry__DerivativeAlreadyRegistered(derivativeIpId);
+        }
+
+        for (uint256 i = 0; i < originalIpIds.length; i++) {
+            if ($.disputeModule.isIpTagged(originalIpIds[i])) {
+                revert Errors.LicenseRegistry__OriginalIpTagged(originalIpIds[i]);
+            }
+            if (derivativeIpId == originalIpIds[i]) {
+                revert Errors.LicenseRegistry__DerivativeIsOriginal(derivativeIpId);
+            }
+            if ($.expireTimes[originalIpIds[i]] < block.timestamp) {
+                revert Errors.LicenseRegistry__OriginalIpExpired(originalIpIds[i]);
+            }
+            // derivativeIp can only register with default license terms or the same license terms as the original IP
+            if ($.defaultLicenseTemplate != licenseTemplate || $.defaultLicenseConfigId != licenseConfigIds[i]) {
+                if ($.licenseTemplates[originalIpIds[i]] != licenseTemplate) {
+                    revert Errors.LicenseRegistry__OriginalIpUnmachedLicenseTemplate(originalIpIds[i], licenseTemplate);
+                }
+                if (!$.attachedLicenseConfigs[originalIpIds[i]].contains(licenseConfigIds[i])) {
+                    revert Errors.LicenseRegistry__OriginalIpHasNoLicenseConfig(originalIpIds[i], licenseConfigIds[i]);
+                }
+            }
+            $.originalIps[derivativeIpId].add(originalIpIds[i]);
```

Comment:
Definitely think the `original` vs `derivative` naming is confusing. If I understand correctly, this code should work for any level in the "IP tree", so original throws me off, is this just setting parent nodes and child nodes?
What if we create a `GraphNode` data struct with the mappings? That might help with terminology.

Reply:
make sense.  renamed to parent and child.